cocos/gfx/webgpu/webgpu-command-buffer.ts

```diff
@@ -0,0 +1,740 @@
+                        passEncoder?.draw(ia.vertexCount, instanceCount, ia.firstVertex, ia.firstInstance);
+                    };
+                    this._renderPassFuncQueue.push(drawFunc);
+                }
+            }
+
+            ++this._numDrawCalls;
+            this._numInstances += inputAssembler.instanceCount;
+            const indexCount = inputAssembler.indexCount || inputAssembler.vertexCount;
+            if (this._curGPUPipelineState) {
+                const glPrimitive = this._curGPUPipelineState.glPrimitive;
+                switch (glPrimitive) {
+                    case 'triangle-strip':
+                        this._numTris += (indexCount - 2) * Math.max(inputAssembler.instanceCount, 1);
+                        break;
+                    case 'triangle-list': {
+                        this._numTris += indexCount / 3 * Math.max(inputAssembler.instanceCount, 1);
+                        break;
+                    }
+                    default:
+                        break;
+                }
+            }
+        } else {
+            console.error('Command \'draw\' must be recorded inside a render pass.');
+        }
+    }
+
+    public updateBuffer(buffer: Buffer, data: BufferSource, offset?: number, size?: number) {
+        if (this._type === CommandBufferType.PRIMARY && !this._isInRenderPass
+            || this._type === CommandBufferType.SECONDARY) {
+            const gpuBuffer = (buffer as unknown as WebGPUBuffer).gpuBuffer;
+            if (gpuBuffer) {
+                const cmd = this._webGPUAllocator!.updateBufferCmdPool.alloc(WebGPUCmdUpdateBuffer);
+                let buffSize = 0;
+                let buff: BufferSource | null = null;
+
+                // TODO: Have to copy to staging buffer first to make this work for the execution is deferred.
+                // But since we are using specialized primary command buffers in WebGL backends, we leave it as is for now
+                if (buffer.usage & BufferUsageBit.INDIRECT) {
+                    buff = data;
+                } else {
+                    if (size !== undefined) {
+                        buffSize = size;
+                    } else {
+                        buffSize = (data as ArrayBuffer).byteLength;
+                    }
+                    buff = data;
+                }
+                const device = WebGPUDeviceManager.instance;
+                const nativeDevice = (device as WebGPUDevice).nativeDevice;
+                nativeDevice?.queue.writeBuffer(gpuBuffer.glBuffer!, gpuBuffer.glOffset, buff as ArrayBuffer);
+
+                // cmd.gpuBuffer = gpuBuffer;
+                // cmd.buffer = buff;
+                // cmd.offset = (offset !== undefined ? offset : 0);
+                // cmd.size = buffSize;
+                // this.cmdPackage.updateBufferCmds.push(cmd);
+                //
+                // this.cmdPackage.cmds.push(WebGPUCmd.UPDATE_BUFFER);
```

Comment:
Need cleanup. I recommend there is no code in comment.
If there is code needed to be kept for some reason, we can add `if (ENABLE_XXX) { ... }` surround the needed code block.